druid/examples/focus.rs

```diff
@@ -0,0 +1,136 @@
+// Copyright 2020 The Druid Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Demonstrates focus and focus scope widgets
+
+pub use druid::theme;
+use druid::widget::prelude::*;
+use druid::widget::{Button, Container, Flex, Focus, FocusScope, Label, TextBox, WidgetExt};
+
+use druid::{AppLauncher, Color, Data, LocalizedString, Point, Rect, WidgetPod, WindowDesc};
+
+pub struct FocusDecorator<T> {
+    pub(crate) child: WidgetPod<T, Box<dyn Widget<T>>>,
+}
+
+/// A container that draws a border around its child when focused."
+impl<T: Data> FocusDecorator<T> {
+    pub fn new(child: impl Widget<T> + 'static) -> Self {
+        FocusDecorator {
+            child: WidgetPod::new(child).boxed(),
+        }
+    }
+}
+
+impl<T: Data> Widget<T> for FocusDecorator<T> {
+    fn event(&mut self, ctx: &mut EventCtx, event: &Event, data: &mut T, env: &Env) {
+        self.child.event(ctx, event, data, env);
+    }
+
+    fn lifecycle(&mut self, ctx: &mut LifeCycleCtx, event: &LifeCycle, data: &T, env: &Env) {
+        if let LifeCycle::HotChanged(_) = event {
+            ctx.request_paint();
+        }
+        self.child.lifecycle(ctx, event, data, env);
+    }
+
+    fn update(&mut self, ctx: &mut UpdateCtx, _old_data: &T, data: &T, env: &Env) {
+        self.child.update(ctx, data, env);
+    }
+
+    fn layout(&mut self, ctx: &mut LayoutCtx, bc: &BoxConstraints, data: &T, env: &Env) -> Size {
+        let size = self.child.layout(ctx, &bc, data, env);
+        let rect = Rect::from_origin_size(Point::ORIGIN, size);
+        self.child.set_layout_rect(ctx, data, env, rect);
+
+        size
+    }
+
+    fn paint(&mut self, ctx: &mut PaintCtx, data: &T, env: &Env) {
+        self.child.paint(ctx, data, &env);
+
+        if ctx.focus_node().is_focused {
```

Comment:
couldn't this still just be `if ctx.is_focused()`? like we have an `is_focused()` method that internally checks the `FocusNode`?

Reply:
Yes, it can be, but I kept `is_focused` unchangeable because `focus_node()` is the current focus widget information. On the other hand, `is_focused` shows if the current widget is focused. I just was not sure if we could change the behavior of the `is_focused` method.